### PR TITLE
ESDT v2 implementations on SystemSC - dynamic tokens and tokenType

### PR DESCRIPTION
**Pull request by sasurobert**
## Reasoning behind the pull request
- 
- 
- 
  
## Proposed changes
- 
- 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: vm/systemSmartContracts/esdt.go [hunk 1]
```diff
@@ -2050,6 +2204,214 @@ func (e *esdt) stopNFTCreateForever(args *vmcommon.ContractCallInput) vmcommon.R
...
+	for _, esdtRole := range token.SpecialRoles {
+		for _, role := range esdtRole.Roles {
+			mapOfRoles[string(role)]++
+		}
+	}
+
+	rolesWithHaveToBeSingular := rolesForDynamicWhichHasToBeSingular()
+	for _, role := range rolesWithHaveToBeSingular {
+		if mapOfRoles[role] > 1 {
+			return vm.ErrCannotChangeToDynamic
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:2344]:
```GO
return fmt.Errorf("%w, role %s was found multiple times", vm.ErrCannotChangeToDynamic, role)
``` 
?

File: vm/systemSmartContracts/esdt.go [hunk 2]
```diff
@@ -2050,6 +2204,214 @@ func (e *esdt) stopNFTCreateForever(args *vmcommon.ContractCallInput) vmcommon.R
...
+		Topics:     [][]byte{args.Arguments[0], token.TokenName, token.TickerName, token.TokenType},
+	}
+	e.eei.AddLogEntry(logEntry)
+
+	e.sendTokenTypeToSystemAccounts(args.CallerAddr, args.Arguments[0], token)
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) sendTokenTypeToSystemAccounts(caller []byte, tokenID []byte, token *ESDTDataV2) {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:2397]: Let's not forget about unit/integration tests for the new code

File: common/enablers/epochFlags.go [hunk 3]
```diff
@@ -757,3 +759,8 @@ func (holder *epochFlagsHolder) NFTStopCreateEnabled() bool {
 func (holder *epochFlagsHolder) IsChangeOwnerAddressCrossShardThroughSCEnabled() bool {
 	return holder.changeOwnerAddressCrossShardThroughSCFlag.IsSet()
 }
+
+// DynamicESDTEnabled return true if the dynamicESDTFlag is enabled
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply common/enablers/epochFlags.go:763]: done

File: vm/systemSmartContracts/esdt.go [hunk 4]
```diff
@@ -43,7 +43,20 @@ const canCreateMultiShard = "canCreateMultiShard"
 const upgradeProperties = "upgradeProperties"
 
 const conversionBase = 10
+
 const metaESDT = "MetaESDT"
+const nonFungibleV2 = "NonFungibleESDTV2"
+const ESDTSetTokenType = "ESDTSetTokenType"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:49]: done

File: vm/systemSmartContracts/esdt.go [hunk 5]
```diff
@@ -1550,16 +1603,75 @@ func (e *esdt) isSpecialRoleValidForNonFungible(argument string) error {
...
 	}
 }
 
+func (e *esdt) isSpecialRoleValidForDynamicSFT(argument string) error {
+	err := e.isSpecialRoleValidForDynamicNFT(argument)
+	if err == nil {
+		return nil
+	}
+
+	switch argument {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:1662]: done

File: vm/systemSmartContracts/esdt.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:2344]: done

File: vm/systemSmartContracts/esdt.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/esdt.go:2397]: doing it.

File: vm/systemSmartContracts/esdt.go [hunk 6]
```diff
@@ -2264,6 +2267,11 @@ func (e *esdt) createDynamicToken(args *vmcommon.ContractCallInput) ([]byte, *ES
 	numOfDecimals := uint32(0)
 	if isWithDecimals {
 		propertiesStart++
+		if len(args.Arguments) < propertiesStart {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:2280]: :+1:

File: vm/systemSmartContracts/esdt.go [hunk 7]
```diff
@@ -378,7 +378,7 @@ func (e *esdt) registerNonFungible(args *vmcommon.ContractCallInput) vmcommon.Re
 	}
 
 	tokenType := []byte(core.NonFungibleESDT)
-	if e.enableEpochsHandler.DynamicESDTEnabled() {
+	if e.enableEpochsHandler.IsFlagEnabled(common.DynamicESDTFlag) {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/esdt.go:381]: since we have merged the PR with the rc/v1.7.0, please add the `common.DynamicESDTFlag` to the list on L126

**Review by BeniaminDrasovean — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Review by BeniaminDrasovean — DISMISSED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**